crates/admin/src/service.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use axum::error_handling::HandleErrorLayer;
+use futures::FutureExt;
+use http::StatusCode;
+use tower::ServiceBuilder;
+
+use restate_meta::MetaHandle;
+use restate_schema_impl::Schemas;
+use restate_storage_query_datafusion::context::QueryContext;
+use tracing::info;
+
+use crate::{rest_api, state, storage_query};
+use crate::{Error, Options};
+
+pub struct AdminService {
+    opts: Options,
+    schemas: Schemas,
+    meta_handle: MetaHandle,
+}
+
+impl AdminService {
+    pub fn new(opts: Options, schemas: Schemas, meta_handle: MetaHandle) -> Self {
+        Self {
+            opts,
+            schemas,
+            meta_handle,
+        }
+    }
+
+    pub async fn run(
+        self,
+        drain: drain::Watch,
+        worker_handle: impl restate_worker_api::Handle + Clone + Send + Sync + 'static,
+        query_context: Option<QueryContext>,
+    ) -> Result<(), Error> {
+        let rest_state =
+            state::AdminServiceState::new(self.meta_handle, self.schemas, worker_handle);
+
+        let router = axum::Router::new();
+
+        // Stitch query http endpoint if enabled
+        let router = if let Some(query_context) = query_context {
+            let query_state = Arc::new(state::QueryServiceState { query_context });
+            // Merge storage query router
+            router.merge(storage_query::create_router(query_state))
+        } else {
+            router
+        };
+
+        let router = router
+            // Merge meta API router
+            .merge(rest_api::create_router(rest_state))
```

Comment:
Can there be the problem that the `rest_api` router tries to register the same path as `storage_query` router?

Reply:
Yes, if this happens the storage_query router wins and overrides. That's unlikely to happen without being noticed to be honest.